Mathlib/SetTheory/Ordinal/Club.lean

```diff
@@ -0,0 +1,285 @@
+/-
+Copyright (c) 2024 Nir Paz. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Nir Paz
+-/
+import Mathlib.SetTheory.Cardinal.Cofinality
+import Mathlib.SetTheory.Ordinal.Topology
+
+/-!
+# Club and stationary sets
+
+This file sets up the basic theory of clubs (closed and unbounded sets) and stationary sets.
+
+## Main definitions
+
+* `Ordinal.IsClosed`: A set of ordinals `S` is closed in `o` if `S ⊆ Iio o`
+  and `S` contains every `x < o` such that `x.IsAcc S`.
+* `Ordinal.IsClub`: A set of ordinals `S` is a club in `o` if
+  it is closed in `o` and unbounded in `o`.
+
+## Main results
+
+* `isClub_sInter`: The intersection of fewer than `o.cof` clubs in `o` is a club in `o`.
+-/
+
+noncomputable section
+
+open Classical Cardinal Set Order Filter
+
+universe u v
+
+namespace Ordinal
+
+/-- A set of ordinals is a club below an ordinal if it is closed and unbounded in it. -/
+def IsClub (C : Set Ordinal) (o : Ordinal) : Prop :=
```

Comment:
Question, does it make sense to have two notions of club sets? One for "club below an ordinal" and also a global one. An example for the latter might be the fixed points of a normal function.

On a similar note, does it make sense to define this for general linear/well orders with the order topology?

Reply:
I assumed we'll just use "the pointwise lift is a club below the universal ordinal" directly for a proper club, but the more I think about it the more annoying that sounds. Things like "the restriction of a universal club at a limit point creates a club below that limit point" would need some universe hopping. So I guess the question is can a definition like this have enough API that we wouldn't have to deal with universes when using it. And I don't know (of course this wouldn't be in this PR anyways). You have experience with universe shenanigans, do you think this would make things more tidy?

I've never heard about clubs over non-wellorders and a quick search didn't yield anything. I don't know if people outside this field ever care a lot if their closed sets in some order topology are also unbounded, to the point that they satisfy many things arbitrary closed sets don't.